src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_template_builder.py

```diff
@@ -472,33 +502,62 @@ def build_vmss_resource(name, naming_prefix, location, tags, overprovision, upgr
         ]
 
     # Build storage profile
-    storage_properties = {
-        'osDisk': {
+    storage_properties = {}
+    if storage_profile in [StorageProfile.SACustomImage, StorageProfile.SAPirImage]:
+        storage_properties['osDisk'] = {
             'name': os_disk_name,
             'caching': storage_caching,
             'createOption': 'FromImage',
         }
-    }
 
-    if storage_profile == StorageProfile.SACustomImage:
-        storage_properties['osDisk'].update({
-            'osType': os_type,
-            'image': {
-                'uri': image
-            }
-        })
-    elif storage_profile == StorageProfile.SAPirImage:
-        storage_properties['osDisk']['vhdContainers'] = "[variables('vhdContainers')]"
-    else:
-        raise CLIError('Unsupported storage profile.')
+        if storage_profile == StorageProfile.SACustomImage:
+            storage_properties['osDisk'].update({
+                'osType': os_type,
+                'image': {
+                    'uri': image
+                }
+            })
+        else:
+            storage_properties['osDisk']['vhdContainers'] = "[variables('vhdContainers')]"
+    elif storage_profile in [StorageProfile.ManagedPirImage, StorageProfile.ManagedPirImage]:
+        storage_properties['osDisk'] = {
+            'createOption': 'FromImage',
+            'caching': storage_caching,
+            'managedDisk': {'storageAccountType': storage_sku}
+        }
 
-    if storage_profile in [StorageProfile.SAPirImage]:
+    if storage_profile in [StorageProfile.SAPirImage, StorageProfile.ManagedPirImage]:
         storage_properties['imageReference'] = {
             'publisher': os_publisher,
             'offer': os_offer,
             'sku': os_sku,
             'version': os_version
         }
+    if storage_profile == StorageProfile.ManagedCustomImage:
+        storage_properties['imageReference'] = {
+            'id': image
+        }
+
+    if data_disk_sizes_gb is not None:
```

Comment:
This is probably what you were talking about, but we should extract this out to a common method if the logic is shared with VM create.